docs/cardano-monads.rst

```diff
@@ -0,0 +1,922 @@
+    ...
+
+This approach mirrors the one of ``mtl``. However, ``mtl`` is concerned with
+general purpose effects, of which there are only a few (under 10, even if we
+consider more exotic ones like reverse state or logic-t), whereas
+domain-specific effects are numerous. Besides, general purpose effects don't pop
+up left and right, whereas domain-specific ones might be created/deleted
+depending on how the application evolves. All in all, this made mirroring
+``mtl`` quite a bad idea.
+
+Pros:
+
+* Simple and familiar. It's easy to reason about effects when there's a 1-1
+  correspondence between classes and transformers. Virtually everyone is
+  familiar with ``mtl`` and can quickly learn to apply this approach.
+
+* Flexibility. Given a monad ``m``, one can easily add an effect to it by
+  putting the relevant transformer on top. It's convenient to use effects
+  locally, not only application-wide.
+
+Cons:
+
+* Bad run-time performance. Having a transformer per effect means no flattening.
+  This boils down to ``ReaderT (a, b, c, d) m`` being more performant than
+  ``ReaderT a (ReaderT b (ReaderT c (ReaderT d m)))`` (same goes for ``StateT``,
+  ``WriterT``, ``ExceptT``, etc). Since transformers for many domain-specific
+  effects are isomorphic to one of these general purpose effects, having a dozen
+  of nested transformers isn't such a good idea. (@int-index has benchmarks to back up
+  this claim).
+
+* No run-time configurability. The stack of transformers determines at
+  compile-time what effect implementations are used. This means that if we
+  wanted to select an implementation at run-time (via a CLI flag, for instance),
+  we had to use a different stack of transformers. The amount of possible
+  combinations grows rather quickly. For instance, if feature X has ``2``
+  implementations and feature Y has ``3``, then we need ``2*3=6`` transformer
+  stacks. (As discovered later, this can be addressed with closed effect sums,
+  but then we lose extensibility, so it's still an issue).
+
+* Cost of introducing new effects. For ``N`` transformers and ``M`` classes,
+  there must be ``N * M - k`` instances (where ``k`` is the amount of unneeded
+  or impossible effect combinations: one does not simply use ``WriterT`` with
+  ``ContT``). The more effects we have in the application, the more costly it is
+  to add one more (at 10 effects we already hit 100 instances, the cost is
+  insurmountable). Not only this is boilerplate-heavy, it's also antimodular, as
+  every effect needs to know about each other.
+
+Conclusion:
+
+* In retrospect, we could probably solve the ``N*M`` issue by careful use of
+  ``{-# OVERLAPPABLE #-}`` instances (although ``Mockable``-related type
+  families would still lead to a fairly large amount of boilerplate).
+  Nevertheless, the run-time performance of highly nested transformer stacks
+  might be unsatisfactory. (Slowdown is linear in the amount of layers).
+
```

Comment:
My 2 cents: I think your initial approach (Using MonadBar for grouping Bar related features etc.) was good. The cons can be easily resolved.

1) As for run-time performance: you forget about monad transformers for a moment. Your monadic code is polymorphic in a monad and uses MonadLog etc. Then you identify which features related to each type class will be globally used (e.g. MonadLog is surely global as you log everywhere, but maybe MonadDB is not global, because you access database only in a specific part of the application).

Then:
- you implement a base monad that provides instances for all type classes considered to be used globally - let's call it Production. It will most likely be a newtype over ReaderT x IO (or StateT x IO), where x is everything needed to provide instances of said type classes. Than solves the performance problem (and reduces boilerplate), because you don't need most of the associated monad transformers you'd write otherwise and your base monad "stack" is flat.

- for each local effect you do the classic approach, i.e. provide an associated transformer (e.g. LocalT for MonadLocal etc.). Then, whenever in you application you need access to these, you simply use runLocalT to extend your stack with needed feature. Example:

```
class MonadLocal m
newtype LocalT t m

runLocalT :: LocalCtx -> LocalT t m a -> m a

foo :: MonadLog m => m ()
foo = do
  ...
  localCtx <- ...
  ...
  runLocalT localCtx bar

bar :: (MonadLog m, MonadLocal m) -> m ()
bar = ...
```

As you already mentioned, you'd also need

```
instance {-# OVERLAPPABLE #-} (MonadTrans t, MonadLocal m) => MonadLocal (t m)
```

so that GHC takes care of lifting for you (this gets rid of cost of introducing new effects).

2) If you need runtime selection, it can be done in a rather straightforward manner: http://lpaste.net/360627

Reply:
> for each local effect you do the classic approach, i.e. provide an associated transformer (e.g. LocalT for MonadLocal etc.). 

This has the same potential for non-transitive capability overrides as manually passing dictionaries. If the base monad has a a `MonadLogger` instance, and another `MonadFoo` instance relies on it, then adding `NoLoggingT` on top will not change the logging behavior in `MonadFoo`.

Reply:
@int-index I'm not sure what the actual use case for that is, but this is not accurate. If you have NoLoggingT in your stack that replaces functions from MonadLog type class with no-ops, then if you run code constrained over MonadFoo using this stack, it will use the no-op MonadLog instance.

Reply:
@arybczak Perhaps we are talking about different issues, but here is a Gist to demonstrate the issue: https://gist.github.com/int-index/fca78d931e59bdd9688054d132d4a241

`runNoLoggingT` did not disable logging completely in the transformer stack, as `foo` continues to log to stdout.

Reply:
Ahh. My bad, for some reason I thought this would work as expected. But of course, if the whole computation is lifted, then yes, that won't work.

I'm still not sure though why would you want to do that.